extensions/standard-processors/processors/AttributesToJSON.cpp

```diff
@@ -0,0 +1,174 @@
+  core::PropertyBuilder::createProperty("Include Core Attributes")
+    ->withDescription("Determines if the FlowFile core attributes which are contained in every FlowFile should be included in the final JSON value generated.")
+    ->isRequired(true)
+    ->withDefaultValue<bool>(true)
+    ->build());
+
+const core::Property AttributesToJSON::NullValue(
+  core::PropertyBuilder::createProperty("Null Value")
+    ->withDescription("If true a non existing or empty attribute will be NULL in the resulting JSON. If false an empty string will be placed in the JSON.")
+    ->isRequired(true)
+    ->withDefaultValue<bool>(false)
+    ->build());
+
+core::Relationship AttributesToJSON::Success("success", "All FlowFiles received are routed to success");
+
+void AttributesToJSON::initialize() {
+  setSupportedProperties({
+    AttributesList,
+    AttributesRegularExpression,
+    Destination,
+    IncludeCoreAttributes,
+    NullValue
+  });
+  setSupportedRelationships({Success});
+}
+
+void AttributesToJSON::onSchedule(core::ProcessContext* context, core::ProcessSessionFactory* /*sessionFactory*/) {
+  std::string value;
+  if (context->getProperty(AttributesList.getName(), value) && !value.empty()) {
+    attribute_list_ = utils::StringUtils::splitAndTrimRemovingEmpty(value, ",");
+  }
+  if (context->getProperty(AttributesRegularExpression.getName(), value) && !value.empty()) {
+    attributes_regular_expression_ = std::regex(value);
+  }
+  write_destination_ = WriteDestination::parse(utils::parsePropertyWithAllowableValuesOrThrow(*context, Destination.getName(), WriteDestination::values()).c_str());
+  context->getProperty(IncludeCoreAttributes.getName(), include_core_attributes_);
+  context->getProperty(NullValue.getName(), null_value_);
+}
+
+bool AttributesToJSON::isCoreAttributeToBeFiltered(const std::string& attribute) const {
+  return !include_core_attributes_ && core_attributes_.find(attribute) != core_attributes_.end();
+}
+
+std::unordered_set<std::string> AttributesToJSON::getAttributesToBeWritten(const std::map<std::string, std::string>& flowfile_attributes) const {
+  std::unordered_set<std::string> attributes;
+
+  for (const auto& attribute : attribute_list_) {
+    if (!isCoreAttributeToBeFiltered(attribute)) {
+      attributes.insert(attribute);
+    }
+  }
+
+  if (attributes_regular_expression_) {
+    for (const auto& [key, value] : flowfile_attributes) {
+      if (!isCoreAttributeToBeFiltered(key) && std::regex_match(key, attributes_regular_expression_.value())) {
+        attributes.insert(key);
+      }
+    }
+  }
+
+  return attributes;
+}
+
+void AttributesToJSON::addAttributeToJson(rapidjson::Document& document, const std::string& key, const std::string& value) {
+  rapidjson::Value json_key(key.c_str(), document.GetAllocator());
+  rapidjson::Value json_val;
+  if (!value.empty() || !null_value_) {
+    json_val.SetString(value.c_str(), document.GetAllocator());
```

Comment:
the nifi implementation (despite what the documentation says) writes `null` only if the attribute does not exist and null value is requested, it does not seem to write `null` when the value of the attribute is an empty string

Reply:
It's again seems to me like a bug on the NiFi side, either in the documentation or in the code. Which one would make more sense?

Reply:
I would support writing `null` only for non-existing attributes, attributes with an empty string value could(should?) have different semantics from non-existing attributes

Reply:
I asked the NiFi team about this and they said the same, the empty string values should not be modified. I'll update according to this.

Reply:
Updated in effc866d914fb3ccff4f70a7a0ebcea6305403fc